app/src/ingestion/markdown_tree.py

```diff
@@ -0,0 +1,602 @@
+            parents[node.data_type].add(node.parent.data_type)
+        if isinstance(node.data, TokenNodeData):
+            tokens[node.data_type].update(node.token.__dict__.keys())
+    return {
+        "counts": counts,
+        "children": children,
+        "parents": parents,
+        "tokens": tokens,
+    }
+
+
+def tokens_vs_tree_mismatches(tree: Tree) -> dict:
+    """
+    Return the tokens' parent-and-children mismatches compared to the tree structure.
+    Use this as a sanity check after manipulating the tree structure.
+    """
+    memo: dict[str, list[str]] = defaultdict(list)
+    for node in tree:
+        if node.data_type == "Document":
+            continue
+        if not isinstance(node.data, TokenNodeData) or not node.is_block_token():
+            continue
+
+        if node.parent:
+            if isinstance(node.parent.data, TokenNodeData):
+                if node.token.parent != node.parent.token:
+                    memo["diff_parent"].append(
+                        f"Different token parent for {node.data_id}: {node.token.parent} vs {node.parent.token}"
+                    )
+        elif node.token.parent:
+            memo["has_parent"].append(f"Token has parent for {node.data_id}: {node.token.parent}")
+
+        if node.children:
+            node_children_tokens = [
+                c.token for c in node.children if isinstance(c.data, TokenNodeData)
+            ]
+            if node_children_tokens != node.token.children:
+                memo["diff_children"].append(
+                    f"Different token children for {node.data_id}: {node_children_tokens} vs {node.token.children}"
+                )
+        elif node.token.children:
+            token_children = [
+                c
+                for c in node.token.children
+                if isinstance(c, block_token.BlockToken) and c.__class__.__name__ != "TableCell"
+            ]
+            if token_children:
+                memo["has_children"].append(
+                    f"Token has block-token children for {node.data_id}: {token_children}"
+                )
+    return memo
+
+
+def render_tree_as_md(tree: Tree, normalize: bool = True) -> str:
+    return render_subtree_as_md(tree.system_root.first_child(), normalize=normalize)
+
+
+def render_subtree_as_md(node: Node, normalize: bool = True) -> str:
+    """
+    Render the node and its descendants (a subtree) to markdown text.
+    Useful for creating markdown text for chunks.
+    Since the structure of the tree (i.e., each node's parent and children) is independent of each Token's parent and children,
+    we cannot rely on mistletoe's renderer (which is based on Tokens) to render the tree correctly. Hence, we have this function.
+    """
+    if node.data_type == "HeadingSection":  # Render the custom HeadingSection node specially
+        out_str = []
+        for c in node.children:
+            out_str.append(render_subtree_as_md(c, normalize=normalize))
```

Comment:
Alternatively, we could customize or create a subclass of MarkdownRenderer (https://github.com/miyuchina/mistletoe/blob/master/dev-guide.md#a-new-renderer) to handle `HeadingSection` but this is easier and more transparent, esp. since we don't need a token to represent HeadingSections.